### PR TITLE
Flink: Add 'cache.expiration-interval-ms' option to FlinkCatalog

### PR DESCRIPTION
**Pull request by lvyanquan**
As what https://github.com/apache/iceberg/pull/3543 has introduced, 'cache.expiration-interval-ms' option allows user to configure caching policy.
But this parameter has not been added in FlinkCatalog yet, This pr is aiming to do that.

### CODE REVIEW

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java [hunk 1]
```diff
@@ -145,8 +145,26 @@ protected Catalog createCatalog(
       baseNamespace = Namespace.of(properties.get(BASE_NAMESPACE).split("\\."));
     }
 
-    boolean cacheEnabled = Boolean.parseBoolean(properties.getOrDefault(CACHE_ENABLED, "true"));
-    return new FlinkCatalog(name, defaultDatabase, baseNamespace, catalogLoader, cacheEnabled);
+    boolean cacheEnabled =
+        Boolean.parseBoolean(properties.getOrDefault(CatalogProperties.CACHE_ENABLED, "true"));
```
Comment (hililiwei) [inline flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:149]: Now that you have introduced `PropertyUtil`, can we unify the style?

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 2]
```diff
@@ -102,14 +102,18 @@ public FlinkCatalog(
...
+      long cacheExpirationIntervalMs) {
     super(catalogName, defaultDatabase);
     this.catalogLoader = catalogLoader;
     this.baseNamespace = baseNamespace;
     this.cacheEnabled = cacheEnabled;
 
     Catalog originalCatalog = catalogLoader.loadCatalog();
-    icebergCatalog = cacheEnabled ? CachingCatalog.wrap(originalCatalog) : originalCatalog;
+    icebergCatalog =
+        cacheEnabled
```
Comment (hililiwei) [inline flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:114]: Can we drop `cacheEnabled` and just use `cacheExpirationIntervalMs`?  When `cacheExpirationIntervalMs` is zero, it is disabled.

Comment (lvyanquan) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:113]: thanks for pointing, address it.

**Comment by lvyanquan** [comment]
I wonder whether should I add doc about it since https://iceberg.apache.org/docs/latest/configuration/#catalog-properties has introduced it.

File: flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java [hunk 3]
```diff
@@ -18,14 +18,13 @@
  */
 package org.apache.iceberg.flink;
 
-import static org.apache.iceberg.flink.FlinkCatalogFactory.CACHE_ENABLED;
```
Comment (stevenzwu) [inline flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java:21]: why this change? is it from spotless formatting?

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java [hunk 4]
```diff
@@ -145,8 +145,14 @@ protected Catalog createCatalog(
...
-    boolean cacheEnabled = Boolean.parseBoolean(properties.getOrDefault(CACHE_ENABLED, "true"));
-    return new FlinkCatalog(name, defaultDatabase, baseNamespace, catalogLoader, cacheEnabled);
+    long cacheExpirationIntervalMs =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+
+    return new FlinkCatalog(
+        name, defaultDatabase, baseNamespace, catalogLoader, cacheExpirationIntervalMs);
```
Comment (stevenzwu) [inline flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:155]: this breaks the compatibility. I feel the benefit probably doesn't justify the breakage. we should keep the boolean config and just support the additional expire interval.

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 2 above)
Comment (stevenzwu) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:113]: this breaks the compatibility. I feel the benefit probably doesn't justify the breakage. we should keep the boolean config and just support the additional expire interval.

**Comment by stevenzwu** [comment]
@hililiwei 's contribution of Flink 1.16 was just merged. @lvyanquan can you also port the change to 1.16 module?

File: flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java (hunk 3 above)
Comment (lvyanquan) [reply flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java:21]: Because CACHE_ENABLED of FlinkCatalogFactory was removed, and this line used string from org.apache.iceberg.CatalogProperties.
```
config.put(CatalogProperties.CACHE_ENABLED, String.valueOf(cacheEnabled));
```

Comment (stevenzwu) [reply flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java:21]: got it

**Comment by lvyanquan** [comment]
Resubmitted to port the change to 1.16 module. 
"cache-enabled" is reserved now, since users who set "cache-enabled" to "false" before would need to add property "cache.expiration-interval-ms = -1" to an existed catalog, It's extra work for users.

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java [hunk 5]
```diff
@@ -145,8 +145,27 @@ protected Catalog createCatalog(
       baseNamespace = Namespace.of(properties.get(BASE_NAMESPACE).split("\\."));
     }
 
-    boolean cacheEnabled = Boolean.parseBoolean(properties.getOrDefault(CACHE_ENABLED, "true"));
-    return new FlinkCatalog(name, defaultDatabase, baseNamespace, catalogLoader, cacheEnabled);
+    boolean cacheEnabled = CatalogProperties.CACHE_ENABLED_DEFAULT;
```
Comment (stevenzwu) [inline flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:148]: we can use `PropertyUtil` here too, right?

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java [hunk 6]
```diff
@@ -145,8 +145,27 @@ protected Catalog createCatalog(
...
+    if (properties.containsKey(CatalogProperties.CACHE_ENABLED)) {
+      cacheEnabled = Boolean.parseBoolean(properties.get(CatalogProperties.CACHE_ENABLED));
+    }
+
+    long cacheExpirationIntervalMs =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+    if (cacheExpirationIntervalMs == 0) {
```
Comment (stevenzwu) [inline flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:158]: we should just do a `Preconditions` check here that this is a positive number.

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java (hunk 5 above)
Comment (lvyanquan) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java]: address it.

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java (hunk 6 above)
Comment (lvyanquan) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: Not sure why limit to positive numbers, since negative values mean "cache expiration is turned off and entries expire only on refresh etc", users can set a negative values.

Comment (hililiwei) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: > * If cache.expiration-interval-ms is zero, caching will be turned off entirely (irrespective of the cache-enabled flag)
> * If cache.expiration-interval-ms is positive (with cache-enabled=true), then caching will be on and entries will expire if not accessed in that period of time.
> * If users pass in a value <= -1, (with cache-enabled=true), then caching will be enabled and cache expiration will never happen (i.e. entries will persist unless explicitly refreshed, original behavior).

detail ref:  https://github.com/apache/iceberg/pull/3543#issue-1052281636

Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: @lvyanquan because we retained the boolean flag for enabling caching, I thought the refresh interval should just be positive. But apparently, I didn't get the implication of negative values. So forget about this comment.

Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: This PR replicates the SparkCatalog behavior, which is fine by me.

> - If cache.expiration-interval-ms is zero, caching will be turned off entirely (irrespective of the cache-enabled flag)
> - If users pass in a value <= -1, (with cache-enabled=true), then caching will be enabled and cache expiration will never happen (i.e. entries will persist unless explicitly refreshed, original behavior).

I find it a little weird that zero and negative numbers have different implications. why distinguish those two? @kbendick can you share some context?

Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: if we look at the code of `CachingCatalog`, it only check positive values. Hence don't understand why we need value zero to disable caching since we already have the boolean flag. `positive` refresh interval would be caching with expiration. otherwise, it is caching without expiration (until refreshed).

```
  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
    Caffeine<Object, Object> cacheBuilder = Caffeine.newBuilder().softValues();

    if (expirationIntervalMillis > 0) {
      return cacheBuilder
          .removalListener(new MetadataTableInvalidatingRemovalListener())
          .executor(Runnable::run) // Makes the callbacks to removal listener synchronous
          .expireAfterAccess(Duration.ofMillis(expirationIntervalMillis))
          .ticker(ticker)
          .build();
    }

    return cacheBuilder.build();
  }
```

Comment (lvyanquan) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: In my opinion, these two parameters actually have some overlap.
If we introduce a value of 0, we actually don't need ‘cache-enabled’, which is exactly what @hililiwei said before.
***
Can we drop cacheEnabled and just use cacheExpirationIntervalMs? When cacheExpirationIntervalMs is zero, it is disabled.https://github.com/apache/iceberg/pull/6111#discussion_r1012530287
***
Here is just for compatibility with the previous configuration.

Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: I understood the overlap btw the boolean flag and the expire value at 0. But dropping `cacheEnabled ` would be a breaking change, which we should try to avoid

Comment (lvyanquan) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: I think the key decision is whether to keep the same [parameter definitions](https://iceberg.apache.org/docs/latest/configuration/#catalog-properties) as before or to remove and verify unnecessary 0 values in Flink. I agree with the same configuration as Spark, because the parameter is public. 
But Flink has a separate [configuration page](https://iceberg.apache.org/docs/latest/flink/#catalog-configuration), so redefining is also acceptable.

Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: > But Flink has a separate [configuration page](https://iceberg.apache.org/docs/latest/flink/#catalog-configuration), so redefining is also acceptable.

I had the same thought. Here is what is intuitive to me.
* if the boolean flag is false, caching is disabled
* if the boolean flag is true
  * if  cacheExpirationIntervalMs > 0, caching automatically expired after the value
  * if <=0, caching never expire automatically

What are you takes? @lvyanquan @hililiwei @pvary

Comment (pvary) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:157]: Coming from a world where we were a platform for multiple companies, I would chose the more conservative approach and try to avoid breaking changes unless it is strictly necessary.

Also adding a new configuration which just provides another way to archive the same function seems something which will lead to confusion down the line, so I would opt for:
- keep the  `cache-enabled` boolean flag
- if the `cache-enabled` is `true` find meaningful values for the `cacheExpirationIntervalMs`, or throw an error message if they do not make sense, so:
   - if `cacheExpirationIntervalMs` is > 0 then expire the the element
   - if `cacheExpirationIntervalMs` is <=0 then never expire the element
   - if `cacheExpirationIntervalMs` is unset then never expire the element to keep the backwards compatibility
- if the `cache-enabled` is `false` then ignore the `cacheExpirationIntervalMs` or throw an exception if it is set.

Just my 2 cents

Comment (lvyanquan) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java]: Addressed for these suggestions:
>   * if `cacheExpirationIntervalMs` is unset then never expire the element to keep the backwards compatibility.    

1. Use CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS_OFF as default value to keep the same behavior.

>   * if <=0, caching never expire automatically. this is the difference. don'y disable caching with value 0 because boolean flag serves the purpose already.

2. Howerver, 0 value is excluded in the constructor of [CachingCatalog](https://github.com/apache/iceberg/blob/dbb8a404f6632a55acb36e949f0e7b84b643cede/core/src/main/java/org/apache/iceberg/CachingCatalog.java) as the following code：
```
  protected CachingCatalog(
      Catalog catalog, boolean caseSensitive, long expirationIntervalMillis, Ticker ticker) {
    Preconditions.checkArgument(
        expirationIntervalMillis != 0,
        "When %s is set to 0, the catalog cache should be disabled. This indicates a bug.",
        CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS);
    this.catalog = catalog;
    this.caseSensitive = caseSensitive;
    this.expirationIntervalMillis = expirationIntervalMillis;
    this.tableCache = createTableCache(ticker);
  }
```
So cacheExpirationIntervalMs is not allowed to be 0, add this restriction to Precondition and Document too.

**Review by stevenzwu — APPROVED** [review]

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java (hunk 6 above)
Comment (stevenzwu) [reply flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java]: latest code looks good to me. Will merge tomorrow unless there are other comments.

We are on the same page regarding backward compatibility for keeping the boolean flag.

> cacheExpirationIntervalMs is not allowed to be 0

agree with the Precondition.  on the other hand, I think it is is a bit unnecessary for the CachingCatalog. it could have treated `<=0` as no expiration.

**Review by hililiwei — APPROVED** [review]
LGTM.

**Comment by stevenzwu** [comment]
thanks @lvyanquan for the contribution and @hililiwei and @pvary for the reviews.
